src/odb.c

```diff
@@ -669,6 +671,97 @@ void git_odb_free(git_odb *db)
 	GIT_REFCOUNT_DEC(db, odb_free);
 }
 
+typedef struct {
+	git_oid oid;
+	git_array_oid_t parents;
+} git_odb__graft;
+
+int git_odb__graft_register(git_odb *odb, const git_oid *oid, git_array_oid_t parents)
+{
+	int res = 0, error = 0;
+	size_t k, i;
+	git_odb__graft *graft;
+	git_oid *parent_oid;
+
+	assert(odb && oid);
+
+	graft = git__calloc(1, sizeof(*graft));
+	GITERR_CHECK_ALLOC(graft);
+
+	git_array_init_to_size(graft->parents, git_array_size(parents));
+	git_array_foreach(parents, i, parent_oid) {
+		git_oid *id = git_array_alloc(graft->parents);
+		GITERR_CHECK_ALLOC(id);
+
+		git_oid_cpy(id, parent_oid);
+	}
+
+	git_oid_cpy(&graft->oid, oid);
+	k = git_oidmap_put(odb->grafts, &graft->oid, &res);
+	if (res < 0) {
+		error = -1;
+		goto cleanup;
+	}
+
+	git_oidmap_set_value_at(odb->grafts, k, graft);
+
+cleanup:
+	return error;
+}
+
+int git_odb__graft_unregister(git_odb *odb, const git_oid *oid)
+{
+	assert(odb && oid);
+
+	git_oidmap_delete(odb->grafts, oid);
+
+	return 0;
+}
+
+static int graft_for_oid(git_odb__graft **out, git_odb *odb, const git_oid *oid)
+{
+	size_t pos;
+
+	assert(out && odb && oid);
+
+	pos = git_oidmap_lookup_index(odb->grafts, oid);
+	if (!git_oidmap_valid_index(odb->grafts, pos)) {
+		return GIT_ENOTFOUND;
+	}
+
+	*out = git_oidmap_value_at(odb->grafts, pos);
+
+	return 0;
+}
+
+int git_odb__graft_object(void *obj, git_odb *odb)
+{
+	git_odb__graft *graft = NULL;
+	git_commit *commit = obj;
+	git_oid *oid;
+	size_t i;
+	int error = 0;
+
+	/* We only handle commit grafts */
+	if (git_object_type((git_object *)commit) != GIT_OBJ_COMMIT)
+		return 0;
+
+	if ((error = graft_for_oid(&graft, odb, git_object_id(&commit->object))) < 0) {
+		return 0;
+	}
+
+	git_array_clear(commit->parent_ids);
+	git_array_init_to_size(commit->parent_ids, git_array_size(graft->parents));
+	git_array_foreach(graft->parents, i, oid) {
+		git_oid *id = git_array_alloc(commit->parent_ids);
+		GITERR_CHECK_ALLOC(id);
+
+		git_oid_cpy(id, oid);
+	}
```

Comment:
Okay, so this is the actual grafting mechanism, I guess.

From a cursory read of git-replace(1) this does not look a 100% correct. What you do here is replace the existing commit's parents in-place. From the man page:

>Create a graft commit. A new commit is created with the same content as <commit> except that its parents will be [<parent>...] instead of <commit>'s parents. A replacement ref is then created to replace <commit> with the newly created commit. See contrib/convert-grafts-to-replace-refs.sh for an example script based on this option that can convert grafts to replace refs.

So instead of editing the existing commit, I think you should instead return a new commit which then replaces the grafted commit. You can convince me otherwise though, I'm not familiar with grafting/replace mechanisms.

Reply:
It's the `cgit` graft approach, replacing parents on object parse.

Reply:
Huh. Personally, I'd think this is the wrong approach. If we just replace in place, we have no mechanism to use those grafts for replaces later on, as that requires us to write the grafted object into the ODB to create the references.